app/src/app/copilot/markdown/react-markdown.tsx

```diff
@@ -0,0 +1,92 @@
+"use client";
+
+import { type FC, memo } from "react";
+import ReactMarkdown, { type Options } from "react-markdown";
+import { CodeBlock } from "./code-block";
+import { cn } from "~/lib/utils";
+
+export const MemoizedReactMarkdown: FC<Options> = memo(
+  ReactMarkdown,
+  (prevProps, nextProps) => prevProps.children === nextProps.children,
+);
+
+export function Markdown({ children }: { children: string }) {
+  return (
+    <MemoizedReactMarkdown
+      className="prose flex-1 dark:prose-invert"
+      components={{
+        code({ className, children, ...props }) {
+          if (Array.isArray(children) && typeof children[0] === "string") {
+            if (children[0] === "▍") {
+              return (
+                <span className="mt-1 animate-pulse cursor-default">▍</span>
+              );
+            }
+
+            children[0] = children[0].replace("`▍`", "▍");
+          }
+
+          const match = /language-(\w+)/.exec(className ?? "");
+
+          return "inline" in props && !props.inline ? (
+            <CodeBlock
+              key={Math.random()}
+              language={match?.[1] ?? ""}
+              value={String(children).replace(/\n$/, "")}
+              {...props}
+            />
+          ) : (
+            // Apply the inline-code class along with any existing classes
+            <code className={cn("inline-code", className)} {...props}>
+              {children}
+            </code>
+          );
+        },
+        table({ children }) {
+          return (
+            <table className="border-collapse border border-black px-3 py-1 dark:border-white">
+              {children}
+            </table>
+          );
+        },
+        th({ children }) {
+          return (
+            <th className="break-words border border-black bg-gray-500 px-3 py-1 text-white dark:border-white">
+              {children}
+            </th>
+          );
+        },
+        td({ children }) {
+          return (
+            <td className="break-words border border-black px-3 py-1 dark:border-white">
+              {children}
+            </td>
+          );
+        },
+        a({ href, children, ...props }) {
+          // Check if the last child is a string and if it ends with a period
+          if (Array.isArray(children) && typeof children.at(-1) === "string") {
+            const lastChild = children.at(-1) as string;
+
+            return (
+              <>
+                <a
+                  href={href}
+                  className="markdown-link text-blue-400"
+                  {...props}
+                >
+                  {children}
+                </a>
+                {!lastChild.trim().endsWith(".") && ". "}
+              </>
+            );
+          }
+
+          return <>{children}</>;
+        },
+      }}
```

Comment:
This doesn't seem to be getting rendered properly. This was to solve a corner case for markdown links in order to add proper punctuation. The link doesn't seem to be getting generated along with the punctuation

Reply:
What is the exact edge case? Could you describe what the link would look like so I could properly extract it and generate the replacement?

Reply:
![Screenshot 2023-12-18 at 1 41 15 PM](https://github.com/near-horizon/horizon/assets/108417131/f866159a-56c3-48d7-baf2-86e741979d9a)